### ISSUE TITLE
Reconcile graceful termination

### ISSUE THREAD

**Issue by Fizzadar** [kubernetes/kubernetes#20497]
I currently have a pod in our cluster that cannot be deleted - it's the only object in it's namespace and was originally created via a Job (which _has_ been removed). It's status is pending and has never been "running".

Issuing `DELETE` requests do nothing and the only logs on the master node are:

```
I0202 13:45:56.587268   23887 event.go:206] Event(api.ObjectReference{Kind:"Job", Namespace:"nicktest", Name:"c2a1bd2f-fce5-4eb4-b79a-e0b63b04fa02", UID:"02b8143f-c9b2-11e5-a763-0401816a1001", APIVersion:"extensions", ResourceVersion:"3588377", FieldPath:""}): reason: 'SuccessfulCreate' Created pod: c2a1bd2f-fce5-4eb4-b79a-e0b63b04fa02-1yfuf
I0202 13:45:56.595291   23948 event.go:206] Event(api.ObjectReference{Kind:"Pod", Namespace:"nicktest", Name:"c2a1bd2f-fce5-4eb4-b79a-e0b63b04fa02-1yfuf", UID:"49007b2e-c9b3-11e5-881e-0401816a1001", APIVersion:"v1", ResourceVersion:"3588406", FieldPath:""}): reason: 'Scheduled' Successfully assigned c2a1bd2f-fce5-4eb4-b79a-e0b63b04fa02-1yfuf to 172.16.25.131
```

Currently this cheeky pod is preventing our build chain from working - is there a way to force remove it? Or, if not, which key would I need to drop in etcd to remove it?

**Comment by Fizzadar** [comment]
Update: actually, I can't delete _any_ pods, running or pending. RC's delete OK as do services, just pods won't budge.

**Comment by jlowdermilk** [comment]
How are you issuing `DELETE` requests? Can you paste the kubectl command you are using as well as the output of `kubectl get pods --namespace=YOUR_NAMESPACE`

**Comment by Fizzadar** [comment]
@jlowdermilk I'm issuing the `DELETE` requests via curl, but the `kubectl` equivilant has the same issue:

```
$ kubectl --namespace nicktest delete pod c2a1bd2f-fce5-4eb4-b79a-e0b63b04fa02-1yfuf
pod "c2a1bd2f-fce5-4eb4-b79a-e0b63b04fa02-1yfuf" deleted
```

And the pod information - until deleted the RC was creating pods over and over as none came online, hence why there are so many!:

```
NAME                                         READY     STATUS        RESTARTS   AGE
c2a1bd2f-fce5-4eb4-b79a-e0b63b04fa02-1yfuf   0/1       Terminating   0          21h
c2a1bd2f-fce5-4eb4-b79a-e0b63b04fa02-qvt4o   0/1       Terminating   0          21h
app-00y8j                             0/2       Terminating   0          21h
app-077th                             0/2       Terminating   0          21h
app-09cyp                             0/2       Terminating   0          21h
app-0b1s9                             0/2       Terminating   0          21h
app-0gwn6                             0/2       Terminating   0          21h
app-0kdhq                             0/2       Terminating   0          21h
app-0kqv4                             0/2       Terminating   0          21h
app-0myn0                             0/2       Terminating   0          21h
app-0ncsw                             0/2       Terminating   0          21h
app-0nptb                             0/2       Terminating   0          21h
app-0of7m                             0/2       Terminating   0          21h
app-0qrvw                             0/2       Terminating   0          21h
app-0u77b                             0/2       Terminating   0          21h
app-0xr73                             0/2       Terminating   0          21h
app-14vpl                             0/2       Terminating   0          21h
app-1bdeq                             0/2       Terminating   0          21h
app-1ecqn                             0/2       Terminating   0          21h
app-1f6xf                             0/2       Terminating   0          21h
app-1obng                             0/2       Terminating   0          21h
app-1obpy                             0/2       Terminating   0          21h
app-1sb5h                             0/2       Terminating   0          21h
app-1sxrj                             0/2       Terminating   0          21h
app-1yobe                             0/2       Terminating   0          21h
app-1zau0                             0/2       Terminating   0          21h
app-202uz                             0/2       Terminating   0          21h
app-208d6                             0/2       Terminating   0          21h
app-20lhm                             0/2       Terminating   0          21h
app-2253g                             0/2       Terminating   0          21h
app-22bwm                             0/2       Terminating   0          21h
app-22w55                             0/2       Terminating   0          21h
app-233cp                             0/2       Terminating   0          21h
app-23dw7                             0/2       Terminating   0          21h
app-25ffh                             0/2       Terminating   0          21h
app-27xtw                             0/2       Terminating   0          21h
app-2belf                             0/2       Terminating   0          21h
app-2chpa                             0/2       Terminating   0          21h
app-2fvjp                             0/2       Terminating   0          21h
app-2hiug                             0/2       Terminating   0          21h
app-2or9u                             0/2       Terminating   0          21h
app-2uln4                             0/2       Terminating   0          21h
app-2xq7u                             0/2       Terminating   0          21h
app-2xx67                             0/2       Terminating   0          21h
app-34fw2                             0/2       Terminating   0          21h
app-365tn                             0/2       Terminating   0          21h
app-3b5tl                             0/2       Terminating   0          21h
app-3c74e                             0/2       Terminating   0          21h
app-3cgt1                             0/2       Terminating   0          21h
app-3gpbh                             0/2       Terminating   0          21h
app-3h8lm                             0/2       Terminating   0          21h
app-3j53v                             0/2       Terminating   0          21h
app-405d5                             0/2       Terminating   0          21h
app-40d8b                             0/2       Terminating   0          21h
app-44m2f                             0/2       Terminating   0          21h
app-453kj                             0/2       Terminating   0          21h
app-48754                             0/2       Terminating   0          21h
app-4cbjv                             0/2       Terminating   0          21h
app-4gdb0                             0/2       Terminating   0          21h
app-4ie6u                             0/2       Terminating   0          21h
app-4karg                             0/2       Terminating   0          21h
app-4nwrb                             0/2       Terminating   0          21h
app-4o6nz                             0/2       Terminating   0          21h
app-4reuk                             0/2       Terminating   0          21h
app-4tlmw                             0/2       Terminating   0          21h
app-4wpqh                             0/2       Terminating   0          21h
app-555fm                             0/2       Terminating   0          21h
app-5a9ga                             0/2       Terminating   0          21h
app-5cr6q                             0/2       Terminating   0          21h
app-5dx3q                             0/2       Terminating   0          21h
app-5f7yv                             0/2       Terminating   0          21h
app-5ffzs                             0/2       Terminating   0          21h
app-5fwb5                             0/2       Terminating   0          21h
app-5hor0                             0/2       Terminating   0          21h
app-5jgqu                             0/2       Terminating   0          21h
app-5p99l                             0/2       Terminating   0          21h
app-5pcbs                             0/2       Terminating   0          21h
app-5q5et                             0/2       Terminating   0          21h
app-5qgt4                             0/2       Terminating   0          21h
app-5s46e                             0/2       Terminating   0          21h
app-5vnki                             0/2       Terminating   0          21h
app-60jfg                             0/2       Terminating   0          21h
app-60lxp                             0/2       Terminating   0          21h
app-66y1q                             0/2       Terminating   0          21h
app-67imr                             0/2       Terminating   0          21h
app-681kg                             0/2       Terminating   0          21h
app-69l75                             0/2       Terminating   0          21h
app-6awdr                             0/2       Terminating   0          21h
app-6cywz                             0/2       Terminating   0          21h
app-6l4yh                             0/2       Terminating   0          21h
app-6qmrw                             0/2       Terminating   0          21h
app-6si4e                             0/2       Terminating   0          21h
app-6wrw5                             0/2       Terminating   0          21h
app-6yqve                             0/2       Terminating   0          21h
app-710r7                             0/2       Terminating   0          21h
app-73f0a                             0/2       Terminating   0          21h
app-746hx                             0/2       Terminating   0          21h
app-7b2np                             0/2       Terminating   0          21h
app-7djja                             0/2       Terminating   0          21h
app-7fbjj                             0/2       Terminating   0          21h
app-7fyzw                             0/2       Terminating   0          21h
app-7iqx9                             0/2       Terminating   0          21h
app-7kbzw                             0/2       Terminating   0          21h
app-7l02w                             0/2       Terminating   0          21h
app-7qrap                             0/2       Terminating   0          21h
app-7xovw                             0/2       Terminating   0          21h
app-7zjku                             0/2       Terminating   0          21h
app-82tej                             0/2       Terminating   0          21h
app-89hqk                             0/2       Terminating   0          21h
app-8ca88                             0/2       Terminating   0          21h
app-8f47e                             0/2       Terminating   0          21h
app-8gmhq                             0/2       Terminating   0          21h
app-8hp91                             0/2       Terminating   0          21h
app-8p3fc                             0/2       Terminating   0          21h
app-8q9oa                             0/2       Terminating   0          21h
app-8qwgo                             0/2       Terminating   0          21h
app-8tgat                             0/2       Terminating   0          21h
app-8xmhr                             0/2       Terminating   0          21h
app-8yiz3                             0/2       Terminating   0          21h
app-8zb4o                             0/2       Terminating   0          21h
app-91to0                             0/2       Terminating   0          21h
app-92bcq                             0/2       Terminating   0          21h
app-94j02                             0/2       Terminating   0          21h
app-98bug                             0/2       Terminating   0          21h
app-9adm7                             0/2       Terminating   0          21h
app-9boxg                             0/2       Terminating   0          21h
app-9i1z2                             0/2       Terminating   0          21h
app-9jo5k                             0/2       Terminating   0          21h
app-9otzb                             0/2       Terminating   0          21h
app-9pj2f                             0/2       Terminating   0          21h
app-9xmop                             0/2       Terminating   0          21h
app-9y7wq                             0/2       Terminating   0          21h
app-9yjb9                             0/2       Terminating   0          21h
app-9z0nz                             0/2       Terminating   0          21h
app-9zn3w                             0/2       Terminating   0          21h
app-a7z5d                             0/2       Terminating   0          21h
app-a86mt                             0/2       Terminating   0          21h
app-acvkm                             0/2       Terminating   0          21h
app-adadg                             0/2       Terminating   0          21h
app-aimes                             0/2       Terminating   0          21h
app-aiuvn                             0/2       Terminating   0          21h
app-al9ho                             0/2       Terminating   0          21h
app-ao5by                             0/2       Terminating   0          21h
app-arcdr                             0/2       Terminating   0          21h
app-azw3g                             0/2       Terminating   0          21h
app-b0o5p                             0/2       Terminating   0          21h
app-b9n97                             0/2       Terminating   0          21h
app-baltb                             0/2       Terminating   0          21h
app-bdx5v                             0/2       Terminating   0          21h
app-bj7u6                             0/2       Terminating   0          21h
app-bu8pc                             0/2       Terminating   0          21h
app-bvd7i                             0/2       Terminating   0          21h
app-bwyga                             0/2       Terminating   0          21h
app-byfyf                             0/2       Terminating   0          21h
app-c37iy                             0/2       Terminating   0          21h
app-c4j9k                             0/2       Terminating   0          21h
app-c5so8                             0/2       Terminating   0          21h
app-c6jii                             0/2       Terminating   0          21h
app-cb7js                             0/2       Terminating   0          21h
app-cb86s                             0/2       Terminating   0          21h
app-cdzjv                             0/2       Terminating   0          21h
app-ce5ym                             0/2       Terminating   0          21h
app-cg8tw                             0/2       Terminating   0          21h
app-ci4fx                             0/2       Terminating   0          21h
app-cihn4                             0/2       Terminating   0          21h
app-crbpq                             0/2       Terminating   0          21h
app-cwckt                             0/2       Terminating   0          21h
app-cwxm6                             0/2       Terminating   0          21h
app-d1ic8                             0/2       Terminating   0          21h
app-d4ii7                             0/2       Terminating   0          21h
app-d7a9c                             0/2       Terminating   0          21h
app-d9683                             0/2       Terminating   0          21h
app-dgxpa                             0/2       Terminating   0          21h
app-dhkbo                             0/2       Terminating   0          21h
app-dn1jk                             0/2       Terminating   0          21h
app-dnhue                             0/2       Terminating   0          21h
app-dxmhm                             0/2       Terminating   0          21h
app-dyb1y                             0/2       Terminating   0          21h
app-e51ne                             0/2       Terminating   0          21h
app-e5s2z                             0/2       Terminating   0          21h
app-e6k3s                             0/2       Terminating   0          21h
app-e89l5                             0/2       Terminating   0          21h
app-e8mmk                             0/2       Terminating   0          21h
app-e9f5q                             0/2       Terminating   0          21h
app-eddxb                             0/2       Terminating   0          21h
app-ee4xr                             0/2       Terminating   0          21h
app-eg1cs                             0/2       Terminating   0          21h
app-eh8s2                             0/2       Terminating   0          21h
app-ein4s                             0/2       Terminating   0          21h
app-eks68                             0/2       Terminating   0          21h
app-elve0                             0/2       Terminating   0          21h
app-emz8r                             0/2       Terminating   0          21h
app-eofty                             0/2       Terminating   0          21h
app-epp7c                             0/2       Terminating   0          21h
app-evy44                             0/2       Terminating   0          21h
app-eyc2r                             0/2       Terminating   0          21h
app-f1rhy                             0/2       Terminating   0          21h
app-f3b2b                             0/2       Terminating   0          21h
app-f524z                             0/2       Terminating   0          21h
app-f7gz4                             0/2       Terminating   0          21h
app-facuq                             0/2       Terminating   0          21h
app-fgw0f                             0/2       Terminating   0          21h
app-fi873                             0/2       Terminating   0          21h
app-fiy9v                             0/2       Terminating   0          21h
app-fkhxd                             0/2       Pending       0          21h
app-flgun                             0/2       Terminating   0          21h
app-fpego                             0/2       Terminating   0          21h
app-fqlcb                             0/2       Terminating   0          21h
app-fygoo                             0/2       Terminating   0          21h
app-g004h                             0/2       Terminating   0          21h
app-g28sb                             0/2       Terminating   0          21h
app-gaw8a                             0/2       Terminating   0          21h
app-gazzs                             0/2       Terminating   0          21h
app-gf7x3                             0/2       Terminating   0          21h
app-gj1df                             0/2       Terminating   0          21h
app-gm1jz                             0/2       Terminating   0          21h
app-go9ly                             0/2       Terminating   0          21h
app-gofog                             0/2       Terminating   0          21h
app-gok2r                             0/2       Terminating   0          21h
app-gunic                             0/2       Terminating   0          21h
app-gx9dp                             0/2       Terminating   0          21h
app-gybgw                             0/2       Terminating   0          21h
app-h1xau                             0/2       Terminating   0          21h
app-h287x                             0/2       Terminating   0          21h
app-h31r4                             0/2       Terminating   0          21h
app-h9339                             0/2       Terminating   0          21h
app-hap1e                             0/2       Terminating   0          21h
app-hbbn9                             0/2       Terminating   0          21h
app-hdjjb                             0/2       Terminating   0          21h
app-hf0wo                             0/2       Terminating   0          21h
app-hn2fj                             0/2       Terminating   0          21h
app-hn2lz                             0/2       Terminating   0          21h
app-hxois                             0/2       Terminating   0          21h
app-hxsr3                             0/2       Terminating   0          21h
app-hz28q                             0/2       Terminating   0          21h
app-hzis1                             0/2       Terminating   0          21h
app-i0fdg                             0/2       Terminating   0          21h
app-i6eyf                             0/2       Terminating   0          21h
app-ia77m                             0/2       Terminating   0          21h
app-id5zp                             0/2       Terminating   0          21h
app-ihwta                             0/2       Terminating   0          21h
app-ihzpp                             0/2       Terminating   0          21h
app-iiais                             0/2       Terminating   0          21h
app-ij9nw                             0/2       Terminating   0          21h
app-ikyl4                             0/2       Terminating   0          21h
app-ipqbi                             0/2       Terminating   0          21h
app-is8am                             0/2       Terminating   0          21h
app-iyhtg                             0/2       Terminating   0          21h
app-j1qon                             0/2       Terminating   0          21h
app-j2el8                             0/2       Terminating   0          21h
app-j9x4q                             0/2       Terminating   0          21h
app-jc7c4                             0/2       Terminating   0          21h
app-jdovy                             0/2       Terminating   0          21h
app-jg31p                             0/2       Terminating   0          21h
app-jh50r                             0/2       Terminating   0          21h
app-jocas                             0/2       Terminating   0          21h
app-jpjul                             0/2       Terminating   0          21h
app-jq1g1                             0/2       Terminating   0          21h
app-jttpd                             0/2       Terminating   0          21h
app-jubmf                             0/2       Terminating   0          21h
app-jvnus                             0/2       Terminating   0          21h
app-jyrp6                             0/2       Terminating   0          21h
app-k4pv8                             0/2       Terminating   0          21h
app-kcky8                             0/2       Terminating   0          21h
app-kcu97                             0/2       Terminating   0          21h
app-kig74                             0/2       Terminating   0          21h
app-kwo5i                             0/2       Terminating   0          21h
app-kyjxo                             0/2       Terminating   0          21h
app-kz3us                             0/2       Terminating   0          21h
app-l0dmn                             0/2       Terminating   0          21h
app-l1c90                             0/2       Terminating   0          21h
app-l2xh3                             0/2       Terminating   0          21h
app-l7j36                             0/2       Terminating   0          21h
app-l8pd5                             0/2       Terminating   0          21h
app-lbmnu                             0/2       Terminating   0          21h
app-lept4                             0/2       Terminating   0          21h
app-ljh2x                             0/2       Terminating   0          21h
app-lm56p                             0/2       Terminating   0          21h
app-lp9b5                             0/2       Terminating   0          21h
app-lp9yr                             0/2       Terminating   0          21h
app-lqpvr                             0/2       Terminating   0          21h
app-ltm7m                             0/2       Terminating   0          21h
app-lv42k                             0/2       Terminating   0          21h
app-lxmhy                             0/2       Terminating   0          21h
app-mb6u4                             0/2       Terminating   0          21h
app-mbjt8                             0/2       Terminating   0          21h
app-md1qj                             0/2       Terminating   0          21h
app-mf081                             0/2       Terminating   0          21h
app-mn0ru                             0/2       Terminating   0          21h
app-mqn9c                             0/2       Terminating   0          21h
app-mt1ww                             0/2       Terminating   0          21h
app-mur2j                             0/2       Terminating   0          21h
app-my2p9                             0/2       Terminating   0          21h
app-n8yzr                             0/2       Terminating   0          21h
app-n8zp1                             0/2       Terminating   0          21h
app-n9ws7                             0/2       Terminating   0          21h
app-nagci                             0/2       Terminating   0          21h
app-nao7d                             0/2       Terminating   0          21h
app-ncz4i                             0/2       Terminating   0          21h
app-nd2bx                             0/2       Terminating   0          21h
app-nu4h4                             0/2       Terminating   0          21h
app-nv3y9                             0/2       Terminating   0          21h
app-nyd4f                             0/2       Terminating   0          21h
app-nz9ed                             0/2       Terminating   0          21h
app-o33pn                             0/2       Terminating   0          21h
app-o3a3f                             0/2       Terminating   0          21h
app-o58b0                             0/2       Terminating   0          21h
app-o6t5g                             0/2       Terminating   0          21h
app-o74dx                             0/2       Terminating   0          21h
app-o9eyh                             0/2       Terminating   0          21h
app-o9nhd                             0/2       Terminating   0          21h
app-oa5q5                             0/2       Terminating   0          21h
app-ogod5                             0/2       Terminating   0          21h
app-oj8e6                             0/2       Terminating   0          21h
app-ojgla                             0/2       Terminating   0          21h
app-olswr                             0/2       Terminating   0          21h
app-or3jh                             0/2       Terminating   0          21h
app-oru6x                             0/2       Terminating   0          21h
app-ouhcl                             0/2       Terminating   0          21h
app-owyvd                             0/2       Terminating   0          21h
app-oxn8v                             0/2       Terminating   0          21h
app-oz8zy                             0/2       Terminating   0          21h
app-p4bx9                             0/2       Terminating   0          21h
app-pba2q                             0/2       Terminating   0          21h
app-pest0                             0/2       Terminating   0          21h
app-pf5ji                             0/2       Terminating   0          21h
app-piabv                             0/2       Terminating   0          21h
app-psh3m                             0/2       Terminating   0          21h
app-pt2ww                             0/2       Terminating   0          21h
app-pu6zy                             0/2       Terminating   0          21h
app-puwug                             0/2       Terminating   0          21h
app-pw6oy                             0/2       Terminating   0          21h
app-pzj2s                             0/2       Terminating   0          21h
app-q21ph                             0/2       Terminating   0          21h
app-qdlva                             0/2       Terminating   0          21h
app-qgc53                             0/2       Terminating   0          21h
app-qh2v4                             0/2       Terminating   0          21h
app-qk64j                             0/2       Terminating   0          21h
app-qoxxt                             0/2       Terminating   0          21h
app-qswf8                             0/2       Terminating   0          21h
app-qtv4l                             0/2       Terminating   0          21h
app-qurw9                             0/2       Terminating   0          21h
app-qxnbd                             0/2       Terminating   0          21h
app-r2iba                             0/2       Terminating   0          21h
app-r3u5q                             0/2       Terminating   0          21h
app-r6j61                             0/2       Terminating   0          21h
app-r71bu                             0/2       Terminating   0          21h
app-r8mr8                             0/2       Terminating   0          21h
app-rbkk7                             0/2       Terminating   0          21h
app-rbo3t                             0/2       Terminating   0          21h
app-rk8xy                             0/2       Terminating   0          21h
app-roqjv                             0/2       Terminating   0          21h
app-rp1p2                             0/2       Terminating   0          21h
app-rskzp                             0/2       Terminating   0          21h
app-ryv5j                             0/2       Terminating   0          21h
app-s4bml                             0/2       Terminating   0          21h
app-s8xpq                             0/2       Terminating   0          21h
app-s9dor                             0/2       Terminating   0          21h
app-sa3dz                             0/2       Terminating   0          21h
app-sa5o1                             0/2       Terminating   0          21h
app-sbgxr                             0/2       Terminating   0          21h
app-sbnn6                             0/2       Terminating   0          21h
app-sdkb3                             0/2       Terminating   0          21h
app-se7sk                             0/2       Terminating   0          21h
app-sqcqm                             0/2       Terminating   0          21h
app-sqjlr                             0/2       Terminating   0          21h
app-stg4v                             0/2       Terminating   0          21h
app-stl0y                             0/2       Terminating   0          21h
app-sulai                             0/2       Terminating   0          21h
app-svzlg                             0/2       Terminating   0          21h
app-sxtya                             0/2       Terminating   0          21h
app-sz8ny                             0/2       Terminating   0          21h
app-t5dwv                             0/2       Terminating   0          21h
app-t6kxe                             2/2       Running       0          21h
app-t9t9u                             0/2       Terminating   0          21h
app-tb64m                             0/2       Terminating   0          21h
app-tfqwb                             0/2       Terminating   0          21h
app-thnw0                             0/2       Terminating   0          21h
app-tjy1z                             0/2       Terminating   0          21h
app-tlqt8                             0/2       Terminating   0          21h
app-tmshy                             0/2       Terminating   0          21h
app-tp8bo                             0/2       Terminating   0          21h
app-tussm                             0/2       Terminating   0          21h
app-tvj2h                             0/2       Terminating   0          21h
app-twy5u                             0/2       Terminating   0          21h
app-tz902                             0/2       Terminating   0          21h
app-u43xa                             0/2       Terminating   0          21h
app-u6afu                             0/2       Terminating   0          21h
app-u7fq5                             0/2       Terminating   0          21h
app-u8t5n                             0/2       Terminating   0          21h
app-uaqiy                             0/2       Terminating   0          21h
app-uatvo                             0/2       Terminating   0          21h
app-uge4p                             0/2       Terminating   0          21h
app-unu8v                             0/2       Terminating   0          21h
app-uoohz                             0/2       Terminating   0          21h
app-urbth                             0/2       Terminating   0          21h
app-uy7ab                             0/2       Terminating   0          21h
app-va8ts                             0/2       Terminating   0          21h
app-vb73s                             0/2       Terminating   0          21h
app-vdgsi                             0/2       Terminating   0          21h
app-vdjfp                             0/2       Terminating   0          21h
app-vf4ok                             0/2       Terminating   0          21h
app-vfx0u                             0/2       Terminating   0          21h
app-vhwsi                             0/2       Terminating   0          21h
app-vhxyx                             0/2       Terminating   0          21h
app-vkfgq                             0/2       Terminating   0          21h
app-vnvdi                             0/2       Terminating   0          21h
app-vvd0a                             0/2       Terminating   0          21h
app-vvi78                             0/2       Terminating   0          21h
app-vyavx                             0/2       Terminating   0          21h
app-w0afb                             0/2       Terminating   0          21h
app-w281q                             0/2       Terminating   0          21h
app-w8jnu                             0/2       Terminating   0          21h
app-watu1                             0/2       Terminating   0          21h
app-wbha3                             0/2       Terminating   0          21h
app-wcl47                             0/2       Terminating   0          21h
app-wcu06                             0/2       Terminating   0          21h
app-wixf5                             0/2       Terminating   0          21h
app-wmxog                             0/2       Terminating   0          21h
app-woi16                             0/2       Terminating   0          21h
app-wpzhw                             0/2       Terminating   0          21h
app-wvuom                             0/2       Terminating   0          21h
app-ww9zu                             0/2       Terminating   0          21h
app-wwdxg                             0/2       Terminating   0          21h
app-wxt2m                             0/2       Terminating   0          21h
app-x06h5                             0/2       Terminating   0          21h
app-x09ul                             0/2       Terminating   0          21h
app-x0w7r                             0/2       Terminating   0          21h
app-x9g10                             0/2       Terminating   0          21h
app-x9upk                             0/2       Terminating   0          21h
app-xbaao                             0/2       Terminating   0          21h
app-xemyv                             0/2       Terminating   0          21h
app-xfoke                             0/2       Terminating   0          21h
app-xgqk5                             0/2       Terminating   0          21h
app-xhqm4                             0/2       Terminating   0          21h
app-xizku                             0/2       Terminating   0          21h
app-xjmxu                             0/2       Terminating   0          21h
app-xlkvg                             0/2       Terminating   0          21h
app-xsh6c                             2/2       Running       0          21h
app-xtnmn                             0/2       Terminating   0          21h
app-y06v7                             0/2       Terminating   0          21h
app-y4qx8                             0/2       Terminating   0          21h
app-y5yet                             0/2       Terminating   0          21h
app-yh866                             0/2       Terminating   0          21h
app-yoxbn                             0/2       Terminating   0          21h
app-yqfrs                             0/2       Terminating   0          21h
app-yu4zh                             0/2       Terminating   0          21h
app-yx4pm                             0/2       Terminating   0          21h
app-yzkn3                             0/2       Terminating   0          21h
app-z299u                             0/2       Terminating   0          21h
app-zd9b6                             0/2       Terminating   0          21h
app-zi5r0                             0/2       Terminating   0          21h
app-ziwcr                             0/2       Terminating   0          21h
app-zno4r                             0/2       Terminating   0          21h
app-zosrj                             0/2       Terminating   0          21h
app-zpz3b                             0/2       Terminating   0          21h
app-zqf6c                             0/2       Terminating   0          21h
app-zt12a                             0/2       Terminating   0          21h
```

**Comment by AdoHe** [comment]
@Fizzadar have you ever tried to restart kubelet? can you provide api-server and kubelet log?

**Comment by Fizzadar** [comment]
@AdoHe I did try a full restart of all kube-* services across the entire cluster, with no luck. Unfortunately we've now re-built our cluster as a quick fix so I no longer have logs access. I did discover a job which potentially created the issue, my working theory:

+ RC w/3 replicas created, all normal
+ Job with superset selectors of the RC created
+ RC sees new job pod and attempts to remove it
+ Job/RC keep adding/deleting each others pods

This still doesn't resolve why I was unable to delete any of the pods, could the above series of events cause that? Pending that I'll close this issue as a one off.

**Comment by Fizzadar** [comment]
Nevermind - the _exact_ same issue has happened again. Except this time the job has completely different labels. So I looked at the logs - there's actually nothing logged when I issue `DELETE` requests to the API - is that normal? It's as if the request never actually reaches the API/scheduler, because nothing happens.

`DELETE` still working for services/rcs so only pods are failing now.

**Comment by jlowdermilk** [comment]
Not seeing `DELETE` in logs is definitely abnormal. If you delete a pod via kubectl, you should see a line in `/var/log/kube-apiserver.log` (on the master) like
```
I0203 18:32:03.309430       5 handlers.go:131] DELETE /api/v1/namespaces/default/pods/nginx2-hgh1a: (5.726494ms) 200 [[kubectl/v1.1.4 (linux/amd64) ...]
```

As for why there are so many pods stuck in terminating, I don't suppose you've set `terminationGracePeriodSeconds` in the podspec for the app pods?

**Comment by Fizzadar** [comment]
Yep - once again I have a replication controller continuously creating pods which never actually provision (815 and counting!) - all pending as usual.

@jlowdermilk I'm not setting `terminationGracePeriodSeconds` - the API is showing it as 30 which I assume is the default.

As for the logs - this I cannot explain: there is _nothing_ in the `kube-apiserver` log since 8am this morning - yet for the last 4 hours I've been creating/deleting pods/rcs/services without any issues. As soon as it gets into a bad state and spam-creates pods the API stops letting me `DELETE` pods - but rcs still work.

I have encountered this three times with three full cluster rebuilds now - every time it appears that pods will no longer be provisioned (the logs show the scheduler assigning them, but it's as if the minions ignore it). I'm still working on the cluster so will try to get some reproducible steps. Unfortunately it's kind of hard when the entire cluster becomes unusable every time it happens!

**Comment by Fizzadar** [comment]
@jlowdermilk so I'm pretty sure this is the API server - it never generates any logs from `handlers.go` - while responding to RC/service deletion. When I restart the service the following logs appear instantly and at that point it stops logging at all:

```
I0205 11:00:00.654342   14215 plugins.go:71] No cloud provider specified.
I0205 11:00:00.654771   14215 master.go:368] Node port range unspecified. Defaulting to 30000-32767.
[restful] 2016/02/05 11:00:00 log.go:30: [restful/swagger] listing is available at https://172.16.19.79:6443/swaggerapi/
[restful] 2016/02/05 11:00:00 log.go:30: [restful/swagger] https://172.16.19.79:6443/swaggerui/ is mapped to folder /swagger-ui/
I0205 11:00:00.758110   14215 server.go:651] Serving insecurely on 0.0.0.0:80
E0205 11:00:01.465093   14215 errors.go:62] apiserver received an error that is not an unversioned.Status: too old resource version: 31475 (31480)
E0205 11:00:01.480262   14215 errors.go:62] apiserver received an error that is not an unversioned.Status: too old resource version: 31475 (31480)
E0205 11:00:01.492251   14215 errors.go:62] apiserver received an error that is not an unversioned.Status: too old resource version: 31475 (31480)
E0205 11:00:01.483089   14215 errors.go:62] apiserver received an error that is not an unversioned.Status: too old resource version: 31475 (31480)
E0205 11:00:01.484394   14215 errors.go:62] apiserver received an error that is not an unversioned.Status: too old resource version: 31475 (31480)
E0205 11:00:01.485762   14215 errors.go:62] apiserver received an error that is not an unversioned.Status: too old resource version: 31475 (31480)
E0205 11:00:01.481774   14215 errors.go:62] apiserver received an error that is not an unversioned.Status: too old resource version: 31475 (31480)
E0205 11:00:01.767051   14215 errors.go:62] apiserver received an error that is not an unversioned.Status: too old resource version: 31475 (31480)
E0205 11:00:01.768159   14215 errors.go:62] apiserver received an error that is not an unversioned.Status: too old resource version: 31475 (31480)
E0205 11:00:01.770917   14215 errors.go:62] apiserver received an error that is not an unversioned.Status: too old resource version: 31475 (31480)
E0205 11:00:01.772604   14215 errors.go:62] apiserver received an error that is not an unversioned.Status: too old resource version: 31475 (31480)
E0205 11:00:01.774219   14215 errors.go:62] apiserver received an error that is not an unversioned.Status: too old resource version: 31475 (31480)
E0205 11:00:01.777101   14215 errors.go:62] apiserver received an error that is not an unversioned.Status: too old resource version: 31475 (31480)
E0205 11:00:01.778083   14215 errors.go:62] apiserver received an error that is not an unversioned.Status: too old resource version: 31475 (31480)
E0205 11:00:01.781086   14215 errors.go:62] apiserver received an error that is not an unversioned.Status: too old resource version: 31479 (31480)
E0205 11:00:01.782389   14215 errors.go:62] apiserver received an error that is not an unversioned.Status: too old resource version: 31475 (31480)
E0205 11:00:01.784329   14215 errors.go:62] apiserver received an error that is not an unversioned.Status: too old resource version: 31475 (31480)
E0205 11:00:01.797740   14215 errors.go:62] apiserver received an error that is not an unversioned.Status: too old resource version: 31475 (31480)
```

Is there anything else which might help debug this?

**Comment by Fizzadar** [comment]
Here's an example of one of the broken/hanging pods:

```
Name:				test-x9ytf
Namespace:			nicktest
Image(s):			docker-registry.edtd.net/myapp-d25d0e6,docker-registry.edtd.net/myapp-d25d0e6,docker-registry.edtd.net/myapp-d25d0e6
Node:				172.16.29.91/
Labels:				commit=d25d0e6,name=myapp,version=latest
Status:				Terminating (expires Thu, 04 Feb 2016 16:20:10 +0000)
Termination Grace Period:	30s
Reason:
Message:
IP:
Replication Controllers:	<none>
Containers:
  nginx:
    Container ID:
    Image:		docker-registry.edtd.net/myapp:latest
    Image ID:
    State:		Waiting
    Ready:		False
    Restart Count:	0
    Environment Variables:
      KUBE:	true
      ENV:	staging
  uwsgi:
    Container ID:
    Image:		docker-registry.edtd.net/myapp:latest
    Image ID:
    State:		Waiting
    Ready:		False
    Restart Count:	0
    Environment Variables:
      KUBE:	true
      ENV:	staging
  redis:
    Container ID:
    Image:		docker-registry.edtd.net/myapp:latest
    Image ID:
    State:		Waiting
    Ready:		False
    Restart Count:	0
    Environment Variables:
      KUBE:	true
      ENV:	staging
No volumes.
No events.
```

**Comment by stepanstipl** [comment]
I'm experiencing same problem with some pods, they'll never get deleted, always stuck in "Terminating" phase. Same with expire value, it's long past it.

I'm running Kubernetes 1.1.7

**Comment by jlowdermilk** [comment]
Can someone from @kubernetes/goog-control-plane take a look?

**Comment by bgrant0607** [comment]
cc @mikedanese

**Comment by davidopp** [comment]
@ncdc @pweil- to find someone to investigate.

**Comment by mikedanese** [comment]
@stepanstipl do your kubelet versions match your master version?

**Comment by ncdc** [comment]
@davidopp I've asked @sjenning to take a look. I'm also looking at a bug in Red Hat's bugzilla that is similar.

**Comment by sjenning** [comment]
@Fizzadar I'm looking for confirmation that I understand the issue.  You are setting up an RC that is spinning up a bunch of pods (greater than the number of replicas you have set I assume), all in Pending state.  When you delete the RC, all the pods go to Terminating, but are not fully removed.  The apiserver continues to be responsive to subsequent requests, although the logging can drop out sometimes.

Is this correct?

**Comment by ncdc** [comment]
@Fizzadar @stepanstipl do either of you have partitions/disks that are 100% full?

**Comment by stepanstipl** [comment]
@mikedanese I'm not 100% whether that was the case when I saw the issue, sorry didn't note that info (I know for a fact I was running 1.1.7 api with 1.1.4 kubectl for a while), now I'm both on 1.1.7 and haven't seen the issue for a while, but to be honest we're changing too many things at once in the moment and pretty much constantly tearing & rebuilding our cluster

@ncdc no, but we did have a lot issues with AWS EBS backed volumes - I believe #18785 issue

**Comment by smarterclayton** [comment]
You can always remove a pod like this by doing `kubectl delete pod NAME --grace-period=0`.  When you have this condition, it's almost always because a node is no longer handling those pods - either the Kubelet no longer thinks it has the same nodeName as before, or the Kubelet can't contact the API server or is unable to make updates.  When you `kubectl describe node 172.16.29.91` do you get results back?

**Comment by ncdc** [comment]
@Fizzadar @stepanstipl is this still happening to you? If so, could you please report the versions of your:

- apiserver(s)
- kubelet(s)
- kubectl

If it's still happening, how easy is it to reproduce?

Thanks!

**Comment by knittingdev** [comment]
We ran into this issue today. It appears we hit a memory limit on our node and docker stopped responding. Once docker stopped responding we were unable to terminate pods. 

Using `kubectl delete pod NAME --grace-period=0` allowed us to delete the pods stuck in the `Terminated` status.

Running Docker 1.8 on CoreOS 815.0.0, 

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"1", GitVersion:"v1.1.7+e4e6878", GitCommit:"e4e6878293a339e4087dae684647c9e53f1cf9f0", GitTreeState:"not a git tree"}
Server Version: version.Info{Major:"1", Minor:"2+", GitVersion:"v1.2.0-alpha.5.675+e21f5974277fe1", GitCommit:"e21f5974277fe1ca7b9a1a976b9050e72cb62bcd", GitTreeState:"clean"}
```

**Comment by mikedanese** [comment]
@knittingdev what is the version of the kubelet?

**Comment by knittingdev** [comment]
@mikedanese `Kubelet Version:        v1.2.0-alpha.5.675+e21f5974277fe1`

**Comment by AdoHe** [comment]
@mikedanese this is strange, docker stopped responding cause node state transfer, the node may be not ok to schedule any new pod, but deleting Terminated pod should be ok.

**Comment by AdoHe** [comment]
@knittingdev could you provide `kubelet` log when you do non-graceful deletion: `kubectl delete pod NAME` ?

**Comment by lavalamp** [comment]
@dchen1107 is it possible that an unresponsive docker might make kubelet not get to the point in the sync loop where it deletes terminated pods?

@davidopp Do we need to build a backup plan into the system (e.g. node controller takes over deletion when kubelet's not doing it)?

@smarterclayton as an alternative backup plan, what do you think about setting a (generous) TTL on pods in TERMINATING?

@knittingdev This is a long shot, but could you also post your kube-controller-manager version? (run it with --version)

**Comment by mikedanese** [comment]
@lavalamp if the docker daemon is down, pods won't be synced, kubelet termination will never complete. But if the node reports unhealthy (which they should if the docker daemon is down) they should be evicted ~7 minutes.

**Comment by mikedanese** [comment]
FYI https://github.com/kubernetes/kubernetes/pull/22100

**Comment by mikedanese** [comment]
I'm not entirely convinced that this is the only issue we are seeing with all these reports though.

**Comment by lavalamp** [comment]
Actually, @mikedanese, I think you're clearly fixing a major source of these reports. It's always possible that there's other similar bugs but I would consider your PR to be fixing this issue.

**Comment by smarterclayton** [comment]
Re: setting TTL we saw significant impact to end to end cluster
latency when using it for delete, so I'd be hesitant.

The node controller should delete a node that stops checking in and
should eventually force delete the pods on that node, so there is
already a mechanism to recover from this that should be in play.  If
the node controller thinks the node is healthy, that's probably a
bigger issue.

**Comment by mikedanese** [comment]
@smarterclayton The node controller did not think the node was healthy. The problem was that the kubelet was not healthy enough to gracefully terminate pods and the eviction path does not handle that.

**Comment by bgrant0607** [comment]
I don't know if this is related, but I just noticed this:
```
[It] should scale a job down
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/job.go:161
STEP: Creating a job
STEP: Ensuring active pods == startParallelism
STEP: scale job down
STEP: Ensuring active pods == endParallelism
[AfterEach] Job
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework.go:84
Feb 26 15:52:34.686: INFO: Waiting up to 1m0s for all nodes to be ready
STEP: Destroying namespace "e2e-tests-job-88kfc" for this suite.
Feb 26 15:52:39.707: INFO: Pod e2e-tests-job-88kfc scale-down-wmgrj on node  remains, has deletion timestamp <nil>
Feb 26 15:52:39.707: FAIL: Couldn't delete ns "e2e-tests-job-88kfc": pods remained within namespace e2e-tests-job-88kfc after deletion: [scale-down-wmgrj]


â€¢ Failure in Spec Teardown (AfterEach) [79.637 seconds]
Job
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/job.go:198
  should scale a job down [AfterEach]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/job.go:161

  Feb 26 15:52:39.707: Couldn't delete ns "e2e-tests-job-88kfc": pods remained within namespace e2e-tests-job-88kfc after deletion: [scale-down-wmgrj]

  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework.go:156
```
kubernetes-jenkins/pr-logs/pull/22078/kubernetes-pull-build-test-e2e-gce/30722

**Comment by davidopp** [comment]
Nice debugging @mikedanese . One thing I always wondered about the graceful deletion implementation is why we do the final deletion of the pod from Kubelet rather than from the master (in the normal case). It seems like this bug wouldn't have happened if we hadn't been relying on kubelet to delete pods when their grace period expires. Also the code would probably be easier to reason about if it weren't distributed between master and kubelet. (Of course kubelet would still need to notice when the grace period starts and expires, so it can do the right things to the pod.)

**Comment by smarterclayton** [comment]
The master doesn't know without the kubelet telling it via some path.
The kubelet always has to write the status of "yes, this has stopped"
before anything on the master can react.  If the kubelet can write
status, it can trigger delete.  If the kubelet is partitioned, you
don't know whether the pod is done or not.

Node controller manages deletion of pods on unhealthy nodes, kubelet
manages deletion of pods on healthy nodes, and we need a controller to
delete pods that aren't on real nodes.

**Comment by davidopp** [comment]
Not trying to redesign the feature, just understand the current design...

Why couldn't it be
1. Server sets DeletionTimestamp
2. Kubelet sends SIGTERM
3. Server deletes pod from etcd when DeletionTimestamp time arrives on server
4. Kubelet sends SIGKILL / removes pod on node when it sees pod has been deleted

IIUC you're saying the way it actually works is
1. Server sets DeletionTimestamp
2. Kubelet sends SIGTERM
3. When DeletionTimestamp arrives on Kubelet, Kubelet sends SIGKILL/removes pod on node, and deletes pod from etcd (via API server of course)

In the first approach the pod never terminates if the kubelet gets permanently partitioned after DeletionTimestamp ... is that the scenario you were trying to avoid?

**Comment by smarterclayton** [comment]
The current design is:

1. Server records deletion timestamp and user provided grace period
seconds (or the default).
2. Kubelet observes the change, sends SIGTERM and Docker applies
sigkill after grace period
3. Caller may shorten grace period at server
4. kubelet observes shortened grace period and applies shorter
TERM-KILL window to docker
5. Kubelet observes container stopped, deletes pod.

The time skew between master and node is unknown, and as a result the
current design allows for the kubelet and the master to be causally
correlated without having synchronized about clocks.  The graceful
period starts from when the kubelet observes the deletion, and is
complete when the pod is deleted.

After gracePeriodSeconds elapses, any component that observes the
deletion request in their own timeline is free to release the pods, as
long as they apply it relative to their own clock.  That ensures the
following does not happen:

1. User deletes with D=30s
2. Network packets lag for 15s
3. Kubelet sees deletion at t=15 with grace 30, sends TERM
4. Master at t=30 sends delete 0
5. Kubelet sends immediate KILL at t=30, this breaching the users
requested interval.

**Comment by stepanstipl** [comment]
@ncdc I believe this issue happened when I killed the AWS EC2 instance that was hosting the pods, so #22100 should fix this

**Comment by justinsb** [comment]
Not entirely sure if this is related, but we do have a known/reported issue where docker stops responding, but the node remains healthy.  See recent comments on #13305 . Apparently `docker ps` will hang, but `docker info` will still return.  Apparently we have recently debated whether to use `ps` or `info` (#20509 and #21347) for the health check.

**Comment by mikedanese** [comment]
That would definitely cause this. When I saw this I had run into the kernel panic in #20096 but I think I remember the node was NotReady (well verify when I have time to reproduce). kube-controller-manager logs in this https://github.com/kubernetes/kubernetes/issues/22025#issuecomment-189521361 seem to indicate that the node controller was trying to evict the node but failing which would indicate a bug in the node controller.

**Comment by kargakis** [comment]
Deleting a deployment is hitting this now @bgrant0607 @janetkuo 
```
[mkargaki@dhcp129-211 kubernetes]$ kc get events
...
...
...
47s       47s       1         other-rs         ReplicaSet                            Normal    SuccessfulDelete   {replicaset-controller }      Deleted pod: nginx-deployment-811039852-4w0q1
47s       47s       1         other-rs         ReplicaSet                            Normal    SuccessfulDelete   {replicaset-controller }      Deleted pod: nginx-deployment-811039852-38sca
47s       47s       1         other-rs         ReplicaSet                            Normal    SuccessfulDelete   {replicaset-controller }      Deleted pod: nginx-deployment-811039852-9nduf

[mkargaki@dhcp129-211 kubernetes]$ kc get pods
NAME                               READY     STATUS        RESTARTS   AGE
nginx-deployment-811039852-38sca   1/1       Terminating   0          50m
nginx-deployment-811039852-4w0q1   1/1       Terminating   0          52m
nginx-deployment-811039852-9nduf   1/1       Terminating   0          50m
nginx-deployment-811039852-i8xg9   1/1       Terminating   0          36m
```
https://github.com/kubernetes/kubernetes/issues/22088 is most probably related to this.

**Comment by mikedanese** [comment]
@kargakis Were the nodes not ready? Can you post a kubelet.log on a machine with a terminating pod and a controller-manager.log?

**Comment by kargakis** [comment]
@mikedanese nope, the node was up and running. Unfortunately I have brought the cluster down since, will report back with those logs once I hit it again.

**Comment by mikedanese** [comment]
Okay I've found three things:

1. Pods controlled by daemonsets will remain terminating forever on nodes that can't run a sync loop making the cascading daemonset delete timeout always.
2. AFAICT, pods that fail to be killed (docker returns an error) will log an error in the kubelet but will not be retried unless the kubelet spec changes. cc @dchen1107 
3. If the process cannot be killed (probably a  kernel bug, I've reproduced two separate bugs on both coreos and ubuntu that cause this) then the pod will likely remain terminating until the node reboots. With the bugs I was able to reproduce, the process continued to use a full cpu and did not free any memory.

If anyone runs into any of these, ping me on slack. I'd love to debug this in the wild to see if there are possibly cases I've missed.

**Comment by Random-Liu** [comment]
@mikedanese If a pod fails to be killed, kubelet will retry to kill it again after 10s. However, there is also backoff mechanics when SyncPod returns error, which may lead the period longer.

**Comment by mikedanese** [comment]
Ok thanks for info. Scratch two then. Here's another interesting one:

In https://github.com/kubernetes/kubernetes/issues/22025#issuecomment-189521361 it also looks like the node is fluctuating between ready and not ready, which causes evictions to be canceled and started again. It looks like this can happen when the container runtime is down. Here's the process:

1. kubelet starts and the container runtime is down.
2. kubelet reports healthy
3. kubelet waits 5 minutes for container runtime
4. kubelet's health check fails
5. supervisor d restarts kubelet
6. goto step 1

cc @dchen1107

**Comment by dchen1107** [comment]
Then looks like there is an issue between 1 and 2, Kubelet shouldn't report healthy at the initial state.

**Comment by mikedanese** [comment]
Ok I will try to reproduce. I'm looking for a reason why a kubelet can fluctuate from ready to not ready. We saw this in https://github.com/kubernetes/kubernetes/issues/19118 due to the unregister_netdevice bug

**Comment by mikedanese** [comment]
It seemed likely since the base sync period for kubelet runtime is 5 minutes (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/runtime.go#L28) which corresponds to the period we were seeing in #19118 but there may be other causes for kubelet to flap ready notready. I'm grasping at straws here until i have a reasonable repro to debug.

**Comment by mikedanese** [comment]
at this point pretty certain there are multiple casues of this symptom. I'm considering adding a new loop in node controller that forcefully deletes pods that have been terminating for much much longer (5x or 10x) than their termination grace period.

Pros:
* "Fixes" all possible cases where pods fail to terminate.
Cons:
* could leak processes and resources on nodes that are unable to kill pods.

**Comment by dchen1107** [comment]
What's your kubelet flags when running into unregister_netdevice issue?

**Comment by bgrant0607** [comment]
Have the containers actually terminated in these cases?

Currently we allow pre-stop hooks to run indefinitely. Changing that would be a breaking behavioral change. Will Kubelet still provide that behavior if the pod has been deleted?

**Comment by mikedanese** [comment]
Maybe related to #21367.

**Comment by mikedanese** [comment]
@bgrant0607 In some cases the container has terminated, in some cases it has not.

> Currently we allow pre-stop hooks to run indefinitely. Changing that would be a breaking behavioral change. Will Kubelet still provide that behavior if the pod has been deleted?

@Random-Liu you know the answer?

**Comment by smarterclayton** [comment]
The hooks can run indefinitely but he container still gets SIGTERM and
kill which should end the container and cleanup children if pid1
exits.  A hanging pid1 prevents that of course.

**Comment by mikedanese** [comment]
Ok, I had a meeting with @kubernetes/goog-control-plane and we came to consensus a pod should not be hard deleted until all cleanup is guaranteed to have occurred (e.g. processes have exited, disks unmounted, resource freed). Thus the only entity in the system that should hard delete a pod (or transition it to a failed/succeeded phase) should be the kubelet. If we agree that this is the issue, some consequences:

1. We should remove the termaintePods loop in the node controller and anything else in the master that performs hard deletes. (we could do this for v1.2)
2. Any issue we run into going forward where a pod is stuck in terminating should be assigned to the node team and fixed on a case by case basis.
3. Deletion with an option of grace period = 0 should not be the proxy to a hard delete. grace period = 0 should follow the path of any other termination notice and the option to hard delete a pod should be explicit. (not for v1.2 and may be an api breaking change).

What do you think?

cc @davidopp @dchen1107 @madhusudancs @erictune @quinton-hoole @mml

**Comment by smarterclayton** [comment]
We talked about this behavior during the initial design but walked
back from it - the compromise reflected in the current design is
intended to reflect a midpoint between the two extremes (master or
kubelet being final arbiter of lifecycle).

There will always be cases where pods with deterministic names are
"special" and need to be recreated, while callers wish to have a wait
representative of the cleanup process.  That's what the current design
reflects - acknowledgement that real termination may not be bounded
but that useful forward progress may still be made.  We deliberately
allow the name of the pod to act as a lock, and the grace period
reflects the lock duration.  This change would prevent that by
offering no client side removal of the lock (also, how do you propose
to differentiate between kubelet delete at 0 and user)?

What advantages do you envision that the proposed change bring?  It
will expose more opportunities for debug of degenerate handling in the
node, but it will mean any such bug (or failure, or down node)
completely blocks progress of client code wishing to use the pod name.
To me, I don't see that as a valuable trade off (especially in light
of the PetSet design which benefits from this trade off).  So at least
at first blush, I'm not really in favor of the change and need some
more arguments.  I am fairly against 3, simply because allowing the
user to indicate they want the name back is a powerful tool.

**Comment by erictune** [comment]
Unless the Pod is completely deleted, then resources might still be held.

Examples:
- a process that could not be terminated might be holding and keeping alive
a lease on a lock from some remote system.
- the existence of the pod could mean that a PD is still mounted and cannot
be mounted elsewhere until the pod is deleted.
- a hostPort could still be in use

The name lock is a way that we allow people to ensure all resources are
freed before trying to make a new Pod.

**Comment by smarterclayton** [comment]
> On Mar 1, 2016, at 6:45 PM, Eric Tune <notifications@github.com> wrote:
>
> Unless the Pod is completely deleted, then resources might still be held.
>
> Examples:
> - a process that could not be terminated might be holding and keeping alive
> a lease on a lock from some remote system.
> - the existence of the pod could mean that a PD is still mounted and cannot
> be mounted elsewhere until the pod is deleted.
> - a hostPort could still be in use
>
> The name lock is a way that we allow people to ensure all resources are
> freed before trying to make a new Pod.


Very few of those can be strongly guaranteed, but in general locks
that can't be broken aren't very good locks.  Someone who wants to
guarantee infinite lock can set an infinite grace period today
(without changing the system).  Someone who wants to bound the wait
sets grace period and is ok with a termination from a node controller
at some multiple.  Making all locks potentially infinite and
unbreakablen is a) frustrating to users and b) overly restrictive.

**Comment by mikedanese** [comment]
> how do you propose to differentiate between kubelet delete at 0 and user

There should be an explicit deletion option to hard delete a pod. GracePeriod = 0 is a valid grace period and the pod should not be hard deleted immediately in this case. It should be deleted once it has been cleaned up.

> What advantages do you envision that the proposed change bring?

Two advantages I see:
1. A pod is removed from the api only if it's has been successfully cleaned up. Most users assume that this is the case today when it's not.
2. Termination becomes the sole responsibility of the kubelet which greatly simplifies this logic and establishes a clear boundary.

> Very few of those can be strongly guaranteed, but in general locks
> that can't be broken aren't very good locks.

A lock that is not freed when it should be indicates a very serious system issue and should not be paved over. It should be addressed. Ignoring permafailing cleanup will have very negative down stream effects (e.g. mysteriously failing pet sets).

Also I'm not saying there shouldn't be in API to hard delete a pod. I'm saying that it should not bed GracePeriod = 0, and that nothing in the system but kubelet should use it.

**Comment by davidopp** [comment]
+1 to all of the points Mike and Eric made. I particularly like Mike's observation that what we're proposing in (3) isn't getting rid of the ability of clients to force-delete a pod via some mechanism, just getting rid of the discontinuity we have today where we handle 0 very differently from all other values of grace period.

I think our experience so far with the hybrid approach has shown that it's confusing (for developers and users) and bug-prone.

**Comment by smarterclayton** [comment]
Maybe I don't have the same perspective, but the confusion I've seen
is almost entirely that calling delete doesn't have any effect when
the node is "broken".  And when they do learn about --grace-period=0,
they invoke it with the explicit desire to make the pod go away *now*.
All of the confusion has resulted from bugs or node problems, not from
actual graceful termination itself.

I don't think anything proposed here improves that, and taking away
immediate deletion on 0 makes the experience worse when you
unequivocally want the pod gone.

Fixing user confusion would probably be a) indicating what the time
period is on termination, b) detecting when the grace period has
expired and doing what the user expects (make the pod go away), and c)
do a better job of explaining that the terminating state is likely an
error state, not a normal state.

I would argue the bug prone area is the kubelet in this case - it
can't fulfill its duties, and we make no allowances for that prior to
node deletion.  I agree we always want to fix cases where the kubelet
is obviously not able to fulfill its responsibility, but the changes
described here don't clearly improve anything except the visibility of
the problem (by giving users no recourse to force delete a pod).

**Comment by smarterclayton** [comment]
On Mar 1, 2016, at 8:06 PM, Mike Danese <notifications@github.com> wrote:

how do you propose to differentiate between kubelet delete at 0 and user

There should be an explicit deletion option to hard delete a pod.
GracePeriod = 0 is a valid grace period and the pod should not be hard
deleted immediately in this case. It should be deleted once it has been
cleaned up.

I'm fine with "force delete", but if the user doesn't specify a grace
period then the default will be used.  Force delete with graceful doesn't
seem to make sense at first glance.

What advantages do you envision that the proposed change bring?

Two advantages I see:
1. A pod is removed from the api only if it's has been successfully cleaned
up. Most users assume that this is the case today when it's not.


Do you mean, by changing the kubelet to properly clean up resources on the
node prior to signaling delete (and that today that is not the case)?  That
wasn't clear from your previous comment.


2. Termination becomes the sole responsibility of the kubelet which greatly
simplifies this logic and establishes a clear boundary.


Termination is already the sole domain of the kubelet - maybe I don't
understand what you mean by that.  force delete still has to be represented
on the API server.


Very few of those can be strongly guaranteed, but in general locks
that can't be broken aren't very good locks.

A lock that is not freed when it should be indicates a very serious system
issue and should not be paved over. It should be addressed. Ignoring
permafailing cleanup will have very negative down stream effects (e.g.
mysteriously failing pet sets).


An unavailable system can be just as bad.  But I'm ok with separation of
force delete from grace period.


Also I'm not saying there shouldn't be in API to hard delete a pod. I'm
saying that it should not bed GracePeriod = 0, and that nothing in the
system but kubelet should use it

I'm ok with that, assuming we can sort through the backwards compat issues
(which are significant).

**Comment by smarterclayton** [comment]
Having gone back to this for a few hours, I walk back a little my
support for separating force deletion and the termination period,
because it allows a new class of object - ghost pods.  These are pods
that are undergoing graceful termination, are *intended* to be on
nodes (if only to get clean shutdown) but for which no record exists
in the API.

That is the primary reason why grace period 0 is required to purge
pods - because the user must explicitly signify they wish to remove
the pod immediately.  If the kubelet restarts, it may terminate those
containers with prejudice upon discovering them no longer present in
the API.

Separating forceful deletion *must* severely constrain the amount of
time pods can remain resident, or a fraction of the cluster is now
occupied by processes the kubelet must provide some guarantee for, but
otherwise cannot reflect back to the API.  Any reasonable definition
of "severely constrain" basically looks like "quickly" - which is what
0 is interpreted to mean in the kubelet today.

If we don't intend to allow indefinite or long termination with force
delete, how does specifying graceful and force benefit a user?  If the
two are not specified together or are not beneficial, or if force
means "delete after X seconds no matter what", then the current system
can accomplish that by enforcing the  invariant in multiple places
(node controller and kubelet, with some leeway accounted for in the
node controller due to latency).

**Comment by davidopp** [comment]
@smarterclayton I don't fully understand your last comment. IIUC you are concerned about pods that are still on Kubelet but have been deleted from etcd by the master. I don't see how the proposal will make that worse. Any time you issue the delete from the master instead of from Kubelet, you can end up in this scenario. We already do it today in some circumstances. 

My understanding of (3) is that we make the 0 case symmetric with the non-0 case (by having Kubelet issue the final delete in all cases), but don't prevent users from directly issuing the same delete to the API server that kubelet does when it has finished deleting a pod. There should be no routine need for using that back-door, because we want the list of pods to reflect what's on kubelet, and kubelet is authoritative as to what is on kubelet. The user should only use it when they're sure that the pod is really gone from Kubelet -- they shouldn't use it to create a scenario where a pod is still on Kubelet and now they're forcefully deleting it from the API, creating a ghost.

**Comment by smarterclayton** [comment]
My point was that forceful delete *should* be logically equivalent to
a grace period of zero, because we do not want to support scenarios
where a pod does not exist, but the kubelet must continue to treat the
pod as being in graceful termination.

Force deleting a pod is saying "I want this gone now".  If you don't
want it gone now, you provide a grace period (I am willing to wait X
seconds for this pod to be gone).  If you do want it "gone now", *all*
components of the system should treat that request as a request to
forcibly terminate / take all measures to end that item (the same as
if the cluster were rolled back to a state where the pod no longer
exists).  If the pod lingers due to cleanup or bugs, that's a fault in
the system, not in the users desire to remove the accounting of that
pod from their quota. The user gives up safe deletion for the
guarantee that the system frees the resources charged to them.

Introducing a "force delete" distinct from "grace period 0" is
redundant.  From the cli, I would absolutely be in favor of "kubectl
delete X --force", but i still believe it to be the equivalent of
--grace-period=0.

**Comment by gmarek** [comment]
IIUC @davidopp is distinguishing two cases: Kubelet is alive, and Kubelet is dead and want to have two separate 'grace period 0' calls for them, as the effects of deleting Pod in those cases are different. If Kubelet is alive and is able to handle 'forceful' deletion request it will e.g. unmount PD from the pod, thus making it reusable. If Kubelet is dead we will indeed release the name of the pod, but some of the resources previous instance of this pod can be still locked (e.g. Kubelet is dead, but the machine itself is healthy and PD is still mounted).

This makes those two meaning of gracefull delete 0 requests slightly different, even though they both mean that "*all* components of the system should treat that request as a request to forcibly terminate / take all measures to end that item" - first one is clean shutdown, when we'll try to unmount stuff, second is unsafe and user should know that before it decides to issue this command.

The difference (only one I see) is that in the case of healthy Kubelet following scenario is possible:
- user deletes pod with graceful termination period = 0
- pod gets deleted from the API server
- user's code get the watch event that pod was deleted and tries to recreate it somewhere (bigger, newer, whatever) using the same PD before Kubelet managed to clean up after the previous pod (because e.g. watch event lagged somewhere)
Note that this scenario is impossible for periods > 0, so users *may* think that it should work for 0 as well. IIUC @mikedanese @erictune and @davidopp would like to add to 'delete' contract that it's guaranteed to remove Pod object only after we're sure that we're not keeping any of it's resources anymore. I'm not sure what current contract for delete is.

I agree that it's great from user call perspective, but we should make sure we can define a NodeController semantics in such a world in a non-confusing way. I.e. we need to make it crystal clear that if NC decides that the Node is unhealthy and it needs to remove all the Pods from it, some of the resources assigned to those pods may be leaked (we probably need to do this in the current world anyway).

**Comment by smarterclayton** [comment]
During the original design discussions, we distinguished the following conditions:

1. User is willing to wait to know the pods has been terminated
2. User is unwilling to wait for the pod to complete terminating

The current implementation removes the ability of the user to have a third outcome, which is intentional (as discussed above).  I think grace-period=1 is functionally identical to the scenario you describe as clean-shutdown (since on the kubelet no matter what grace period you provide the minimum termination window is actually 2s).  I agree that from a user confusion perspective allowing two types of grace period zero may have benefits in the CLI, although it's close enough to 1 that I'd rather just describe the difference between 0 and 1 in the API more clearly and have the CLI give humans affordances.

We should probably have a VC to discuss the other implications here.  There's obviously some strong opinions about this.

**Comment by smarterclayton** [comment]
Opened https://github.com/kubernetes/kubernetes/pull/23756 to add `--force` to delete which sets --grace-period=0 and is more intuitive to user expectation.  I think this simplifies the experience for an admin trying to resolve a broken Kubelet.

**Comment by ianmiell** [comment]
We are experiencing a similar issue in production, and are obviating it by restarting the docker service, which appears to resolve.

Although previously we had noted that docker ps was not responding, in the latest incidents it was.

Do we have a set of steps to aid debugging/help identify the root cause that I can pass to our run team?

**Comment by macb** [comment]
Feel like I might be running into this as well but I'm unsure after reading most of the comments above.
```
$ kubectl get po
pufferfish-2747628150-hek8m                 1/1       Terminating        0          3d        ipaddr
```

After bumping up the verbosity of logs and restarting `kubelet` I see:
```
kubelet: I0812 13:17:00.017310    9852 manager.go:1371] Killing container "8bfe02e26c4e67ef862e273cdebaa55280eff21d25a7a36e208b3d214c613b4e pufferfish default/pufferfish-2747628150-hek8m" with 30 second grace period
kubelet: I0812 13:17:00.048009    9852 kubelet.go:2414] SyncLoop (PLEG): "pufferfish-2747628150-hek8m_default(7b1c7181-5da0-11e6-bae9-04013c998d03)", event: &pleg.PodLifecycleEvent{ID:"7b1c7181-5da0-11e6-bae9-04013c998d03", Type:"ContainerStarted", Data:"8bfe02e26c4e67ef862e273cdebaa55280eff21d25a7a36e208b3d214c613b4e"}
kubelet: I0812 13:17:00.052100    9852 kubelet.go:2414] SyncLoop (PLEG): "pufferfish-2747628150-hek8m_default(7b1c7181-5da0-11e6-bae9-04013c998d03)", event: &pleg.PodLifecycleEvent{ID:"7b1c7181-5da0-11e6-bae9-04013c998d03", Type:"ContainerStarted", Data:"8bb548f5ae0164663e209d7c0207e309d059f6e3ee4a4af385afc4cfe29acb29"}
kubelet: W0812 13:17:01.298031    9852 request.go:627] Throttling request took 184.031887ms, request: GET:https://kube-apiserver:6443/api/v1/namespaces/default/pods/pufferfish-2747628150-hek8m
kubelet: W0812 13:17:01.498029    9852 request.go:627] Throttling request took 185.534848ms, request: PUT:https://kube-apiserver:6443/api/v1/namespaces/default/pods/pufferfish-2747628150-hek8m/status
```

Nothing more about that pod in the logs after that. Is this the same behavior or should I open a new issue?

```
$ kubectl delete po pufferfish-2747628150-hek8m --grace-period=0
```
Resulted in an actual deletion of the pod as expected.

**Comment by smarterclayton** [comment]
Now covered by proposal https://github.com/kubernetes/kubernetes/pull/34160

**Comment by dvohra** [comment]
Got same issue: memory limit was hit and deleting the rc does not delete the Pods, which continue to be in Terminate state. The following command deletes the "terminating" pods.
`kubectl delete pods --all --grace-period=0`

**Comment by bgrant0607** [comment]
I believe this is obsolete.
